src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -533,5 +538,154 @@ Json::Value Profile::ToJson() const
+        virtual ~VariablePair() = default;
+        std::wstring rawValue;
+        std::wstring resolvedValue;
+    };
+
+    // Function Description:
+    // - Resolves a key's value from within envMap. Uses recursion to recursively resolve
+    //   referenced keys. envMap is updated as values are resolved so we don't double up on
+    //   work
+    // Arguments:
+    // - key: the key that may be in envMap or the environment (or unknown)
+    // - envMap: the map with all the keys, raw and resolved values
+    // - seenKeys: the list of keys that we have already seen when attempting to resolve the
+    //             current chain of keys. seenKeys needs to be passed in as a parameter so
+    //             that each layer in the recursion only has keys seen leading up to it
+    // Return Value:
+    // - The resolved value for key
+    std::wstring ResolveEnvironmentVariableValue(const std::wstring& key, std::map<std::wstring, VariablePair>& envMap, std::list<std::wstring> seenKeys = {})
+    {
+        // First check if the key is in our environment variable map
+        auto it = envMap.find(key);
+        if (it != envMap.end())
+        {
+            if (!it->second.resolvedValue.empty())
+            {
+                // key has already been resolved, so just return it
+                return it->second.resolvedValue;
+            }
+            if (std::find(seenKeys.cbegin(), seenKeys.cend(), key) != seenKeys.end())
+            {
+                // If we've attempted to resolve this key already then we're circularly resolving keys and this is
+                // an error, not aborting here would result in a dead-lock. e.g.:
+                // {
+                //    'KEY_1': '${env:KEY_2}',
+                //    'KEY_2': '${env:KEY_1}'
+                // }
+                std::wstringstream error;
+                error << L"Self referencing keys in environment settings: ";
+                for (auto seenKey : seenKeys)
+                {
+                    error << L"'" << seenKey << L"', ";
+                }
+                error << L"'" << key << L"'";
+                throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+            }
+            static const std::wregex parameterRegex(LR"(\$\{env:(.*?)\})");
+            auto& value = it->second;
+            auto textIter = value.rawValue.cbegin();
+            std::wsmatch regexMatch;
+            while (std::regex_search(textIter, value.rawValue.cend(), regexMatch, parameterRegex))
+            {
+                if (regexMatch.size() != 2)
+                {
+                    std::wstringstream error;
+                    error << L"Unexpected regex match count '" << regexMatch.size()
+                          << L"' (expected '2') when matching '" << std::wstring(textIter, value.rawValue.cend()) << L"'";
+                    throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));

```

Comment:
Perhaps that should be a fail-fast exception if it can only happen when `parameterRegex` is incorrectly defined.

Reply:
@zadjii-msft, what do you think about this? Should we `RaiseFailFastException` and terminate under this (and other) unexpected error conditions?

Reply:
I always lean towards not fail-fasting if at all possible. This seems like a perfectly fine way to gracefully handle the error for now.